### PR TITLE
Remove Snakemake from Collagen IV segmentation workflow

### PR DESCRIPTION
**Pull request by mfs4rd**
# Description
This PR removes the use of snakemake to parallelize the segmentation and post-processing of Collagen IV results. This is to make the workflow compatible with users without access to a slurm cluster. In addition, the cyto-dl configs have been updated to be compatible with the latest version and a script is provided to generate the csv file needed to run segmentations in this manner. Additionally, optimizations have been made to the post-processing script to make it more efficient in processing segmentation results sequentially without changing the underlying methods.

# Changes
- Remove snakemake as a dependency for the workflow
- Updated cyto-dl configs to utilize using a single czi file as the input for segmenting multiple scenes
- Providing a script for generating a csv file to use for segmenting from a single czi file with cytodl
- Optimizations to post-processing script which increase processing speed by 3x
- Replace `aicsimageio` with `bioio` for reading/writing images
- Changing relevant dependencies in `pyproject.toml` and `pdm.lock`
- Changes to `README.md` file to reflect changes in workflow

### CODE REVIEW

File: CollagenIV_segmentation/README.md [hunk 1]
```diff
@@ -1,42 +1,39 @@
 The workflow for generating basement membrane is divided into three parts.
 
-# Part 1: Create snakemake configs and split out basement membrane channel and timepoints and 
+# Part 1: Create csv file for running segmentations
```
Comment (vianamp) [inline CollagenIV_segmentation/README.md:3]: Should we explain a little more the content of this CSV file?

File: CollagenIV_segmentation/README.md [hunk 2]
```diff
@@ -1,42 +1,39 @@
...
 ```
 
 # Part 2: CytoDL basement membrane segmentation
 
-The data, model, and experiment config are found in the cytoDL_configs directory. The model weights that we are using for the segmentation model can be found here:
+The data, model, and experiment config are found in the `cytoDL_configs` directory. The model weights that we are using for the segmentation model can be found here:
 
-```
+```bash
 /allen/aics/assay-dev/computational/data/EMT_deliverable_processing/cytodl_experiments/logs/train/runs/basement_membrane_semseg/basement_membrane_semseg_version_6_early_model/2023-12-15_16-09-07/checkpoints/epoch_478.ckpt
```
Comment (vianamp) [inline CollagenIV_segmentation/README.md]: This should not use AICS hardcoded paths.

File: CollagenIV_segmentation/cytoDL_configs/data/segmentation_basement_membrane_from_czi.yaml [hunk 3]
```diff
@@ -1,9 +1,23 @@
 
-  _target_: cyto_dl.datamodules.czi.make_CZI_dataloader 
-  csv_path: ${paths.data_dir}/test.csv
-  channel_column: Collagen_channel
+_target_: monai.data.DataLoader
+num_workers: 16
+batch_size: 1
+dataset:
+  _target_: cyto_dl.datamodules.multidim_image.MultiDimImageDataset
+  csv_path: /allen/aics/users/filip.sluzewski/Projects/basement-membrane-segmentation/data_manifests/lumenoid-thickness.csv
```
Comment (vianamp) [inline CollagenIV_segmentation/cytoDL_configs/data/segmentation_basement_membrane_from_czi.yaml]: No AICS paths.

File: CollagenIV_segmentation/cytoDL_configs/experiment/segmentation_basement_membrane.yaml [hunk 4]
```diff
@@ -14,26 +14,22 @@ defaults:
...
+target_col: out_path
+column_channel: channel
+scene_col: scene
 spatial_dims: 3
 raw_im_channels: 1
 
-trainer:
-  max_epochs: 500
+paths:
+  output_dir: /allen/aics/assay-dev/computational/data/EMT_deliverable_processing/Collagen_segmentation_segmentations/membrane_thickness_proof_of_concept
```
Comment (vianamp) [inline CollagenIV_segmentation/cytoDL_configs/experiment/segmentation_basement_membrane.yaml]: No AICS paths

File: CollagenIV_segmentation/cytoDL_configs/experiment/segmentation_basement_membrane.yaml [hunk 5]
```diff
@@ -14,26 +14,22 @@ defaults:
...
+
+checkpoint:
+  ckpt_path: /allen/aics/assay-dev/computational/data/EMT_deliverable_processing/cytodl_experiments/logs/train/runs/basement_membrane_semseg/basement_membrane_semseg_version_6_early_model/2023-12-15_16-09-07/checkpoints/epoch_478.ckpt
```
Comment (vianamp) [inline CollagenIV_segmentation/cytoDL_configs/experiment/segmentation_basement_membrane.yaml]: No AICS paths.

Comment (mfs4rd) [reply CollagenIV_segmentation/cytoDL_configs/experiment/segmentation_basement_membrane.yaml]: Are we going to be providing the model checkpoint on quilt? If it's already up where can I find it so that I can write instructions for downloading it.

**Comment by mfs4rd** [comment]
@vianamp Regarding removing AICS paths from the code, should I simply put in generic paths or comments?

File: CollagenIV_segmentation/README.md (hunk 1 above)
Comment (mfs4rd) [reply CollagenIV_segmentation/README.md:3]: I can add some explanation of the csv file's format

File: CollagenIV_segmentation/cytoDL_configs/experiment/segmentation_basement_membrane.yaml (hunk 5 above)
Comment (vianamp) [reply CollagenIV_segmentation/cytoDL_configs/experiment/segmentation_basement_membrane.yaml]: I believe we did. @smishra3 can you confirm and give @mfs4rd the S3 path?

**Comment by vianamp** [comment]
@smishra3 please help @mfs4rd to deal with AICS paths to conform with the rest of the repo.

File: CollagenIV_segmentation/README.md (hunk 2 above)
Comment (smishra3) [reply CollagenIV_segmentation/README.md]: This can be modified to direct the user to download the checkpoint file from s3 (similar to ACM, either by clicking or using curl) from this path https://open.quiltdata.com/b/allencell/tree/aics/emt_timelapse_dataset/supplemental_files/cytodl_checkpoints/collagenIV_mask_seg_model_checkpoint.ckpt.

File: CollagenIV_segmentation/cytoDL_configs/experiment/segmentation_basement_membrane.yaml (hunk 5 above)
Comment (smishra3) [reply CollagenIV_segmentation/cytoDL_configs/experiment/segmentation_basement_membrane.yaml]: This can be modified to direct the user to download the checkpoint file from s3 (similar to ACM, either by clicking or using curl) from this path https://open.quiltdata.com/b/allencell/tree/aics/emt_timelapse_dataset/supplemental_files/cytodl_checkpoints/collagenIV_mask_seg_model_checkpoint.ckpt.

File: CollagenIV_segmentation/README.md (hunk 1 above)
Comment (smishra3) [reply CollagenIV_segmentation/README.md:3]: I'm not very sure if we need an automatic way to generate the data csv. For all cells mask an example csv format is shown in README and we ask users to follow it for your own data. Further, the prepare_cytodl_csv.py requires multiple user inputs (some contain default values) that might lead to confusion.

File: CollagenIV_segmentation/cytoDL_configs/experiment/segmentation_basement_membrane.yaml (hunk 4 above)
Comment (smishra3) [reply CollagenIV_segmentation/cytoDL_configs/experiment/segmentation_basement_membrane.yaml]: Management of relative paths (for passing cytoDL configs and output directory generation) can be performed using a similar approach that we have used for ACM. You can take a look at this branch - sm_colony_mask_pred_test_v0.

File: CollagenIV_segmentation/README.md (hunk 1 above)
Comment (mfs4rd) [reply CollagenIV_segmentation/README.md:3]: I included this as much for our own internal use as for others. I needed to generate a csv like this when working on generating some BM segmentations for Leigh last week and so it will help for speed and reproducibility to have a ready-made script to generate the csv's we need. Also I don't think that it's confusing to ask users to provide paths to files they want to process as inputs to a script they need to run, we do it everywhere in both EMT repo's.

File: CollagenIV_segmentation/README.md (hunk 2 above)
Comment (mfs4rd) [reply CollagenIV_segmentation/README.md]: Added this in a recent commit

File: CollagenIV_segmentation/cytoDL_configs/data/segmentation_basement_membrane_from_czi.yaml (hunk 3 above)
Comment (mfs4rd) [reply CollagenIV_segmentation/cytoDL_configs/data/segmentation_basement_membrane_from_czi.yaml]: fixed

**Comment by mfs4rd** [comment]
I've implemented the requested changes, as well as added a script which can be used to prepare segmentations for use in the mesh generation workflow. @vianamp @smishra3 please review the changes and let me know if everything is ready for merging now.

File: CollagenIV_segmentation/README.md (hunk 2 above)
Comment (vianamp) [reply CollagenIV_segmentation/README.md]: Let's test against one example you have.

**Review by vianamp — DISMISSED** [review]
Just test the checkpoint and ready to merge.

**Comment by mfs4rd** [comment]
I've tested it on the quilt model weights and everything looks good. @smishra3 can you submit a review so that the branch is published?

**Review by vianamp — APPROVED** [review]
Do you need to test the model again now that the path has changed?

**Comment by mfs4rd** [comment]
No, this is just correcting how you download the checkpoint through curl rather than going to the quilt page and downloading it manually. It's the same model weights tested.

**Review by smishra3 — APPROVED** [review]
